### ISSUE TITLE
Google Captcha

### ISSUE THREAD

**Issue by sut12** [searx/searx#729]
Hey,

i have the problem that google gives me captchas. Therefor i cant see google results. My Server is not blacklistet anywhere. So im wondering what i can do to get rid of the captchas. My instance is not hosted under https. Is that maybe a reason for the google captchas?

Thanks!

**Comment by dalf** [comment]
If you can, use a browser using the same IP than your server : google will throw some captchas at first, and will usually stop after few requests.
One way to do it : install Firefox on the server, then use ssh with X forwarding or VNC.

Not sure, if you need to check if your IP is blacklisted ( how to check : http://www.dnsbl.info/dnsbl-database-check.php not sure either that this is right service to  use)

**Comment by sut12** [comment]
Hello,
i used elinks. It is possible to save images. So i saved the captcha and downloaded it afterwards. For now it is working but i had to enter the captcha already more than once.

I am not listed on any blacklist except on spamcannibal:
generic/anonymous/un-named IP

I need to check how i can get rid of this message. But im not sure if it has something to do with the google problem.

**Comment by sut12** [comment]
It is so wired. Yesterday evening it worked after i enterd the captchas. This morning it didnt work. 

Now it works but i didnt enter any captchas since yesterday...

**Comment by cy8aer** [comment]
Google is dead for me (as startpage).

**Comment by sut12** [comment]
I need google. But its working sometimes and sometimes not. I even tried to analyze with google webmaster tools...

Right now its working - 2morrow maybe not.

**Comment by cy8aer** [comment]
Seems that google sets GOOGLE_ABUSE_EXEMPTION cookie after captcha. And this cookie has a timeout date (sometimes). If there would be some proxying mechanism for the captcha for re-setting the cookie would be nice

**Comment by cy8aer** [comment]
what about forwarding the captcha to the frontend and save the cookie in the backend then?

**Comment by sut12** [comment]
I think that would be the best solution.

**Comment by ghost** [comment]
I wonder how searx.me(and other searx admin) handle google captcha.
Sure they process 1K+/day request to google without getting "scroogled".

**Comment by gszathmari** [comment]
I am having the same issue. It doesn't help if I solve the CAPTCHAs from a browser as the `GOOGLE_ABUSE_EXEMPTION` should be sent with my searx queries.

**Comment by kvch** [comment]
@gszathmari What do you mean? What is `GOOGLE_ABUSE_EXEMPTION`?

**Comment by gszathmari** [comment]
This is the cookie what Google sets if you solve the CAPTCHA. 

So even though I solve the CAPTCHA from the same IP address as my searx instance, searx still fails to retrieve the results from Google. This is because searx is not sending in the `GOOGLE_ABUSE_EXEMPTION` cookie with the queries, so it is still presented by the CATPCHA. However, my browser is not asked to solve the CAPTCHA because it is sending in the `GOOGLE_ABUSE_EXEMPTION` cookies with the subsequent queries.

Hope this clarified up a bit

**Comment by sut12** [comment]
Thats exactly the point. cy8aer also pointed in the right direction: 

> what about forwarding the captcha to the frontend and save the cookie in the backend then?

I also tried to admin my domain with google webmaster tools. So that goole doenst ask for captcha anymore. It didnt help.

**Comment by cy8aer** [comment]
Yeah but I did not know the name of the cookie 'GOOGLE_ABUSE_EXEMPTION'. Thanks to gszathmari

**Comment by prolibre** [comment]
**Engines cannot retrieve results:
google (unexpected crash: CAPTCHA required)** 

every day or almost google blocks me... it's getting unmanageable for me.

Anthony

**Comment by zwnk** [comment]
i recently updated to 0.13.1 and get the same error. went back to 0.12 and get google results.

**Comment by sut12** [comment]
hmmm. I am on 0.12.0 and dont get google results. Maybe i should try to update xD.
Oh wait - rightn ow im getting results. But i am sure tomorrow it wont work again.

**Comment by kvch** [comment]
@zwnk have you tried the latest master? there were a few changes since the last release which fixed  known google issues.

**Comment by zwnk** [comment]
i pulled today and had the google captcha error.

**Comment by Pofilo** [comment]
This is not the first person asking for the problem with google.
As it is solved with a commit since the last release, maybe we can do a new release ?

**Comment by asciimoo** [comment]
@Pofilo we can release a new version (0.13.2), but as I see these fixes don't solve the problem permanently for everybody.

**Comment by sut12** [comment]
I was on 12.0 and it just worked for 3 days now. Today in the morning it didnt work again. No results from googel. So i did an update to 13.1. Now it works again. Changed nothing else. Lets see how long it works.

**Comment by Dominion0815** [comment]
same problem here with 0.13.1:

```
ERROR:searx.search:engine google : exception : CAPTCHA required
Traceback (most recent call last):
  File "/usr/local/searx/searx/search.py", line 104, in search_one_request_safe
    search_results = search_one_request(engine, query, request_params)
  File "/usr/local/searx/searx/search.py", line 87, in search_one_request
    return engine.response(response)
  File "/usr/local/searx/searx/engines/google.py", line 217, in response
    raise RuntimeWarning(gettext('CAPTCHA required'))
RuntimeWarning: CAPTCHA required

```

**Comment by sachaz** [comment]
Hi,

I just updated https://searx.aquilenet.fr to 13.1 and I still got the same issue as Dominion0815

**Comment by sachaz** [comment]
issue corrected, my instance was bugged by bots. Corrected with filtron:
https://asciimoo.github.io/searx/admin/filtron.html
https://github.com/asciimoo/filtron

**Comment by Pofilo** [comment]
Interesting, I will install filtron and give a feedback !

**Comment by steckerhalter** [comment]
I'm using filtron but still get the captcha. Any tips maybe on configuring filtron or something?

**Comment by dalf** [comment]
@steckerhalter : just in case, are you sure that the ```X-Forwarded-For``` header is set when the requests are forwarded to Filtron?

**Comment by steckerhalter** [comment]
@dalf ah maybe not, I'm using Apache. So let me check that, thanks.

**Comment by sachaz** [comment]
here is a link on what I made to make it work (for me):
https://atelier.aquilenet.fr/projects/services/wiki/Searx

**Comment by steckerhalter** [comment]
@sachaz thanks I'll try that

**Comment by borysn** [comment]
there seems to be a lot of mixed feedback whether filtron helps or not, can someone help me setup filtron for my instance?

i am not understanding how rules.json is supposed to work, and i thought the sample config was just going to be an easy drop in, but its not. and i have no idea what to set any of those values to

**Comment by steckerhalter** [comment]
for the record, I have configured filtron but google would still demand the captcha sometimes. the search is used by quite a few people. in the end we decided to disable google altogether.

**Comment by Pofilo** [comment]
@miicha thanks for the feedback !

How did you add the cookies into google.py ? Can you give us a copy of the portion of the file ? _(hidding all sensible values of course !!)_

It seems we have a solution (I have some doubts about how long it will works, maybe the duration of the cookie), but it is not a solution we can fix with a PR. If we all have the same cookies, I think google will block them.

**Comment by miicha** [comment]
It seems to be enough to add these 3 lines (for example before: `params['url'] = search_url.format(offset=offset,`):
`params['cookies']['_ga'] = 'xxxxxxxxxxxx' # 2 years | Used to distinguish users.`
`params['cookies']['_gid'] = 'xxxxxxxxxxxxxx' # 24 hours | Used to distinguish users.`
`params['cookies']['1P_JAR'] = '2018-6-13-17' # changes hourly`
info about lifetime according to https://developers.google.com/analytics/devguides/collection/analyticsjs/cookie-usage and my own observation
maybe even 1 or 2 of these lines are sufficient...

**Comment by asciimoo** [comment]
This solution raises privacy questions. Session cookies can be used to build a user profile. It would be good to find a solution which doesn't increase the sent information about the instance.

**Comment by sachaz** [comment]
google (unexpected crash: CAPTCHA required) 
is back again on our instance after some 4 months of happiness with Filtron. :(

**Comment by dadosch** [comment]
what about using temporary ipv6 addresses for google? Would that help, so that google can't flag us because of many requests per IP (is it??)

**Comment by steckerhalter** [comment]
I'd say the solution is simple: just don't use google. alternatively you can enable startpage which gives you the same results. it's just not worth it trying to satisfy the big g algos.

**Comment by miicha** [comment]
As I wrote earlier, I usee filtron, but still got the google captcha crash.
I configured filtron according to some tutorial I found online (most probably https://asciimoo.github.io/searx/admin/filtron.html and other) but did not check if it was working. Now I found time to check it, and guess what, it wasn't working. Additionally, I turned on logging to investigate what was going on. 

I found massive bot traffic on my domain, searching mostly for download stuff, requesting json output.
After removing ```"filters": ["Param:q", "Path=^(/|/search)$"],``` from my filtron config file, it was finally working and I could turn of the logging again. 

Unfortunately I could not figure out, why the abovementioned filter was not working as intended, but I don't care to much.

If somebody is interested in the rules.json file I could post it somewhere (maybe in the filtron repo?)

**Comment by GitHubGeek** [comment]
Could it be Google tweaking their abuse algorithms to be super stringent? My instance is 100% private (login required) and I'm the only user. Still, the same error comes up after a handful of search.

**Comment by mohe2015** [comment]
@GitHubGeek If you are using a webhoster It could also be that Google blocks a whole ip-range because somebody in there is a spammer. For me that was the case for a mail server.

**Comment by unixfox** [comment]
Hello searx public instance owners,

Since the end of May I have been working on an antibot solution to stop being blocked by Google because of the bad bots that try to do some ranking manipulation on my public instance.

Today I'm sharing with you an early version of the program. It is available here: https://github.com/unixfox/antibot-proxy
You can try it and giving me some feedback on the github repository (please open an issue instead of replying here if you find any problem). If you are currently using filtron you may remove it and use exclusively antibot-proxy for the protection between your main web server and the searx program because antibot-proxy is enough to remove around 99% of the bots.

The program have been running on my personal public searx instance for almost two months: https://searx.be and my searx instance have been almost never been blocked by Google since I deployed my antibot solution. It is not completely without bugs but should work just fine for almost every users.

**Comment by sachaz** [comment]
Yeah ! Thanks unixfox, this is a good idea !
Should we use it above Filtron ?

**Comment by unixfox** [comment]
@sachaz no it's meant to be a better replacement of Filtron.

**Comment by Nosirus** [comment]
how to integrate it into the official docker image?

**Comment by cy8aer** [comment]
Hey, they found a solution with invidious: They now do queries via QUIC! 
https://github.com/omarroth/invidious/issues/811
Is there a useable QUIC implementation for python?

**Comment by unixfox** [comment]
@cy8aer According to Google itself it seems that a QUIC library for Python exist: https://github.com/aiortc/aioquic
~~But looking deeply into the commit mentioned in the issue on invidious repo there is a difference between what invidious and searx request from Google. Invidious in the commit use an API whereas with Searx it's a plain HTML page.
So I don't think Google will render differently their pages based on the protocol.
This needs to be investigated further if that would be really the case.~~
**EDIT**: My bad I didn't see the latest diff in the commit because according to Github it was too large to generate.
Anyway I'm adding that to my task list, it should be very interesting if it works but I think it will probably be quickly fixed by Google engineers. So in the end a temporary fix but not a permanent one.

EDIT²: I just came across a library that has a requests-compatible API with request and the developers are considering adding HTTP3/QUIC to their library: https://github.com/encode/httpx/issues/275 so this means that if httpx gets QUIC support we would technically just need to replace request by httpx in the searx code.

**Comment by bjo81** [comment]
After moving a VM to a different provider, I'm getting always the captcha error. But for my astonishement, opening google,com via elinks does not show a captcha as a user on the server.

**Comment by unixfox** [comment]
It seems that at first google is serving a captcha only for a specific user agent. But then if you try to randomly change the user agent on the same IP all the requests that comes from the same IP will get a captcha even after changing the user agent.

**Comment by bjo81** [comment]
> It seems that at first google is serving a captcha only for a specific user agent. But then if you try to randomly change the user agent on the same IP all the requests that comes from the same IP will get a captcha even after changing the user agent.

Moving the instance to another provider again (away from Hetzner) and it runs fine (until the UI change now)

**Comment by return42** [comment]
> It seems that at first google is serving a captcha only for a specific user agent. But then if you try to randomly change the user agent on the same IP all the requests that comes from the same IP will get a captcha even after changing the user agent.

Update: does not work ..

my 2cent solution:

1. Merged [PR1628](https://github.com/asciimoo/searx/pull/1628)into my [dm-cloud](https://github.com/return42/searx/tree/dm-cloud) branch.

2. To get rid of the Captcha, logged in via ```ssh -X```, installed FFox on my searx server (normally you dont want) .. started FFox, used google search, closed FFox.  Bingo! .. no more CAPTCHA.

Simple, but seems to work (at least for me)

Hopefully using the new google UI (PR1628) no longer triggers captcha .. Thats why I need some testers, penetrating my instance with (google) queries / thanks!

https://darmarit.cloud/searx/

**Comment by unixfox** [comment]
I also tested this method a while ago and it didn't work. It surprised me that it worked for you.
Instead of launching a full Firefox directly on the server you could use a browser automation framework like [puppeteer](https://github.com/puppeteer/puppeteer), just don't forget to make it look like a real Google Chrome browser by using this plugin [puppeteer-extra-plugin-stealth](https://www.npmjs.com/package/puppeteer-extra-plugin-stealth).

I'm still convinced that Google will at some point block the IP if it detects too much bad requests from your server.

Adding the instance to the public list is the way to get quite a lot of requests from bots. It still amazes me that since I launched my antibot project on my Searx public instance I got almost 1800 bad IP that does bad requests:
```
searx:/jail# ls -l | wc -l
1847
```

**Comment by return42** [comment]
> I'm still convinced that Google will at some point block the IP if it detects too much bad requests from your server.

Sure? .. what about enterprises where 2000 and more user use the same IP (with proxy, dns blockers and and and ..).  Never seen that those IP has been blocked.

Thats why I hope it will work :)

**Comment by return42** [comment]
Aargh .. got CAPTCHA back .. you are rigth.

**Comment by unixfox** [comment]
Just use any (free) VPN and make a search on Google. You will immediately get a captcha.
I don't know how big enterprise handle that but the facts are here, Google will present a captcha if too much (bad) request from the same IP are made.

**Comment by unixfox** [comment]
When reading the issues of Invidious, I came across this issue: https://github.com/omarroth/invidious/issues/886.
Basically they suggested that using a solution like `anti-captcha.com` would be a good way to stop the Google recaptcha problem and according to the author it seems that the solution is quite cheap.
Would that be a feasible solution that could be implemented inside Searx or is it too unethical for maintainers to implement a solution like this (behind this service is thousands of workers of poor countries that are working for them)?

**Comment by return42** [comment]
FWIW to mention: I studied my Apache logs and see that bots are queering from tor nodes .. and I also see some queries from https://searx.neocities.org

**Comment by unixfox** [comment]
How can this help/useful for this issue? There are legitimate use for tor.
And it's just so easy to find a big list of proxies.
We won't ban IPs based on the fact that it's just from tor or a VPN provider. It's discrimination.
I already know that bots comes from tor or cheap VPS, I just have to check my list of IP banned for that.

**Comment by return42** [comment]
> How can this help/useful for this issue?

Not really, that is why I labeled it with FWIW.

I am at the beginning .. first I analyzed where queries are come from. Now I see that there are legitimate queries from tor and others nodes.

> We won't ban IPs based 

Correct, we need to ban bots, not because they are all evil, because they are not end-users and searx instances are for the privacy of the end-users.

Google search service is also for end-users and does block bots .. may we need also a captcha or any other solution that is able to distinct between a query from a bot and a human query.

That's why I started to study your anti-bot solution as a first step. I think it goes in the right direction .. I think we need something similar in searx builtin.

For someone who has been involved with it for a long time (like you), all this is a matter of course.

I am at the beginning and I need to sort out my thoughts first. Even if it is a bit verbose, I will do it under this issue, also  to clarify it within the community.  BTW you comments are very useful for / thanks!

**Comment by return42** [comment]
> Basically they suggested that using a solution like anti-captcha.com would be a good way to stop the Google recaptcha problem.

Back to what I said (thought) before:

> We need to ban bots, because they are not end-users and searx instances are for the privacy of the end-users.

I think it is the wrong way to go, to solve CAPTCHA problems for such queries from bots.

My conclusion: _Instead of routing traffic  from bots, we have to avoid traffic from bots._

But: The queries from  https://searx.neocities.org make me also think about, that redirected queries and (may be in the future) clustering searx instances are also valid use cases which can be considered.

**Comment by unixfox** [comment]
No problem, don't hesitate to email me (the address is on my github profile) if you have any question about that.

I completely agree with you that we need to follow the same behavior as Google. 

It would be very difficult to implement and maintaining an antibot feature into Searx itself and I think it's better to have an external solution like filtron or antibot-proxy.

**Comment by dalf** [comment]
A recipe that may work:
* make sure you have a bot filter (filtron, antibot-proxy)
* check the server IP(s) doesn't belong to blacklist : https://whatismyipaddress.com/blacklist-check
* try to tell that your server IP belong to a human (that's the voodoo part) : 
   * setup a proxy / VPN / whatever on the server. Purpose: use a real Firefox to send request from the server IP.
   * start Firefox
   * check that Firefox is using the IP address from your searx instance.
   * create an incognito session.
   * search for something using google.
   * solve the CAPTCHA.
   * close the window.
   * start again : create an incognito session, etc... 
   * I'm not sure when to stop.
* wait for about 2 days (you don't need an oven).
* check if the google engine is working again.

I've done that twice on my instance: the google engine was not working for a long time, and it seems to have solve the issue. I can't say if there is a causal link.

**Comment by unixfox** [comment]
The firefox part is probably unrelated to the actual unblock of Google, I think Google clear their whitelist after a while.
@return42 tried the same thing as you and it stopped working after a few days.

**Comment by return42** [comment]
> @return42 tried the same thing as you and it stopped working after a few days.

Your are talking about https://github.com/asciimoo/searx/pull/1771 ? .. in the meantime I gave up this solution.

> A recipe that may work:
...
> The firefox part is probably unrelated to the actual unblock of Google, I

This is what I observed, the block will be released after a while, after there are no requests from your IP.

**Comment by unixfox** [comment]
I wonder if https://github.com/asciimoo/searx/issues/1785 temporally decreased the amount of bots that used to fetch from the wiki.

**Comment by omarroth** [comment]
> @cy8aer According to Google itself it seems that a QUIC library for Python exist: https://github.com/aiortc/aioquic

Keep in mind that Google does not use IETF-QUIC (indicated by `draft-xx`). They are still developing QUIC internally, and maintain their own versions (indicated by `Q0xx`). You can see this in the `alt-svc` header for Google domains:
```
alt-svc: quic=":443"; ma=2592000; v="46,43",h3-Q050=":443"; ma=2592000,h3-Q049=":443"; ma=2592000,h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000
```

So any library that supports any of: `Q050`, `Q049`, `Q048`, `Q046`, `Q043`. You can see a list of implementations [here](https://github.com/quicwg/base-drafts/wiki/Implementations).

**Comment by unixfox** [comment]
Google is restricting the amount of queries made with HTTP1.1. In HTTP2 there is seems to have less limitation.
As you can see an HTTP 1.1 request with CURL gets blocked:
```
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>302 Moved</TITLE></HEAD><BODY>
<H1>302 Moved</H1>
The document has moved
<A HREF="https://www.google.com/sorry/index?continue=https://www.google.com/search%3Fq%3Dtest%26start%3D0%26gws_rd%3Dcr%26gbv%3D1%26lr%3Dlang_en%26hl%3Den%26ei%3Dx&amp;hl=en&amp;q=EgSygPoDGOD2u_IFIhkA8aeDS8u3MposoeiKX3yV3fq5xmRxo4SnMgFy">here</A>.
</BODY></HTML>
root@vps:~# curl --http1.1 -I -A "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36" "https://www.google.com/search?q=test&start=0&gws_rd=cr&gbv=1&lr=lang_en&hl=en&ei=x"
HTTP/1.1 403 Forbidden
Content-Type: text/html; charset=UTF-8
Date: Thu, 20 Feb 2020 21:37:01 GMT
Server: gws
Content-Length: 1662
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
Alt-Svc: quic=":443"; ma=2592000; v="46,43",h3-Q050=":443"; ma=2592000,h3-Q049=":443"; ma=2592000,h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000
```

But not in HTTP2:
```
<!doctype html><html lang="en-NL"><head><meta charset="UTF-8"><meta content="origin" name="referrer"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>test - Google Search</title><script nonce="MOW9kGL38jsnwjCu4Jh05g==">(function(){
document.documentElement.addEventListener("submit",function(b){var a;if(a=b.target){var c=a.getAttribute("data-submitfalse");a="1"==c||"q"==c&&!a.elements.q.value?!0:!1}else a=!1;a&&(b.preventDefault(),b.stopPropagation())},!0);document.documentElement.addEventListener("click",function(b){var a;a:{for(a=b.target;a&&a!=document.documentElement;a=a.parentElement)if("A"==a.tagName){a="1"==a.getAttribute("data-nohref");break a}a=!1}a&&b.preventDefault()},!0);}).call(this);(function(){
var a=window.performance;window.start=(new Date).getTime();a:{var b=window;if(a){var c=a.timing;if(c){var d=c.navigationStart,f=c.responseStart;if(f>d&&f<=window.start){window.start=f;b.wsrt=f-d;break a}}a.now&&(b.wsrt=Math.floor(a.now()))}}window.google=window.google||{};var h=function(g){g&&g.target.setAttribute("data-iml",(new Date).getTime())};document.documentElement.addEventListener("load",h,!0);google.rglh=function(){document.documentElement.removeEventListener("load",h,!0)};}).call(this);(function(){var gbvu='/search?q\x3dtest\x26lr\x3dlang_en\x26hl\x3den\x26tbs\x3dlr:lang_1en\x26gbv\x3d2\x26sei\x3dBvtOXrHTNtKcsAferKKoDw';(function(){
var e=function(c,a){this.a=c===b&&a||"";this.g=d};e.prototype.i=!0;e.prototype.h=function(){return this.a.toString()};var f=/^(?:(?:https?|mailto|ftp):|[^:/?#]*(?:[/?#]|$))/i,d={},b={};var g,h="1";if(document&&document.getElementById)if("undefined"!=typeof XMLHttpRequest)h="2";else if("undefined"!=typeof ActiveXObject){var k,l,m=["MSXML2.XMLHTTP.6.0","MSXML2.XMLHTTP.3.0","MSXML2.XMLHTTP","Microsoft.XMLHTTP"];for(k=0;l=m[k++];)try{new ActiveXObject(l);h="2";break}catch(c){}}g=h;if("2"==g&&-1==location.search.indexOf("&gbv=2")){var n=gbvu,p=document.getElementById("gbv");p&&(p.value=g);n&&window.setTimeout(function(){var c=location;if(n instanceof e)var a=n;else a=n,a instanceof e||(a="object"==typeof a&&a.i?a.h():String(a),f.test(a)||(a="about:invalid#zClosurez"),a=new e(b,a));c.href=a instanceof e&&a.constructor===e&&a.g===d?a.a:"type_error:SafeUrl"},0)};}).call(this);})();</script><style>body{margin:0 auto;max-width:736px;padding:0 8px}a{color:#1967D2;text-decoration:none;-webkit-tap-highlight-color:rgba(0,0,0,.1)}a:visited{color:#4B11A8}a:hover{text-decoration:underline}img{border:0}html{font-family:Roboto,HelveticaNeue,Arial,sans-serif;font-size:14px;line-height:20px;-webkit-text-size-adjust:100%;color:#3C4043;word-wrap:break-word;background-color:#fff}.bRsWnc{background-color:#fff;border-top:1px solid #e0e0e0;height:39px;overflow:hidden}.N6RWV{height:51px;-webkit-overflow-scrolling:touch;overflow-x:auto;overflow-y:hidden}.Uv67qb{-webkit-box-pack:justify;font-size:12px;line-height:37px;justify-content:space-between;justify-content:space-between}.Uv67qb a,.Uv67qb span{color:#757575;display:block;flex:none;padding:0 16px;text-align:center;text-transform:uppercase;}span.OXXup{border-bottom:2px solid #4285f4;color:#4285f4;font-weight:bold}a.eZt8xd:visited{color:#757575}.FElbsf{border-left:1px solid rgba(0,0,0,.12)}header article{overflow:visible}.Pg70bf{height:39px;display:-webkit-box;display:flex;display:flex;width:100%}.H0PQec{position:relative;flex:1}.sbc{display:flex;width:100%}.Pg70bf input{margin:2px 4px 2px 8px;}.x{width:26px;color:#757575;font:27px/38px arial, sans-serif;line-height:40px;}#qdClwb{flex:0 0 auto;width:39px;height:39px;border-bottom:0;padding:0;border-top-right-radius:8px;background-color:#3b78e7;border:1px solid #3367d6;background-image:url('data:image/svg+xml,\00003csvg height="37px" width="37px" xmlns="http://www.w3.org/2000/svg">\00003cg fill="none" stroke="%23fff">\00003cellipse cx="16.5" cy="16" rx="5.4" ry="5.4" stroke-width="2px">\00003c/ellipse>\00003cpath d="M22 21.3l4.3 4.3" stroke-linecap="round" stroke-width="2px">\00003c/path>\00003c/g>\00003c/svg>');}.sc{font-size:;position:absolute;top:39px;left:0;right:0;box-shadow:0px 2px 5px rgba(0,0,0,.2);z-index:2;background-color:#fff}.sc>div{padding:10px 10px;padding-left:16px;padding-left:14px;border-top:1px solid #DFE1E5}.scs{background-color:#f5f5f5;}.noHIxc{display:block;font-size:16px;padding:0 0 0 8px;flex:1;height:35px;outline:none;border:none;width:100%;-webkit-tap-highlight-color:rgba(0,0,0,0);overflow:hidden;}.sbc input[type=text]{background:none}.sml .cOl4Id{display:none}.l{display:none}.l svg{margin-top:9px}.sml header{background:none}.sml .l{display:block;padding:0 8px}.bz1lBb{background:#fff;border-radius:8px 8px 0 0;box-shadow:0 1px 6px rgba(32, 33, 36, 0.18);margin-top:10px}.KP7LCb{border-radius:0 0 8px 8px;box-shadow:0 2px 3px rgba(32, 33, 36, 0.18);margin-bottom:10px;overflow:hidden}.cOl4Id{text-align:center;padding:14px 0 3px 0;height:35px}.S591j{height:100%}footer{text-align:center;margin-top:18px}footer a,footer a:visited,.smiUbb{color:#5f6368}.ksTU4c{margin:0 13px}#mCljob{margin-top:36px}#mCljob>div{margin:20px}</style></head><body jsmodel=" TvHxbe"><header id="hdr"><script nonce="MOW9kGL38jsnwjCu4Jh05g==">(function(){var c=500;(function(){window.screen&&window.screen.width<=c&&window.screen.height<=c&&document.getElementById("hdr").classList.add("sml");}).call(this);})();</script><div class="cOl4Id"><a href="/?sa=X&amp;ved=0ahUKEwix7veuiuHnAhVSDuwKHV6WCPUQOwgC"><svg stroke="#4285f4" aria-label="Google" fill="none" height="36" stroke-width="18" viewbox="0 0 520 174" width="92"><path d="M69 70.4l61 0m-9 3.5a56 57 0 1 1-17-47" stroke-width="19px"></path><ellipse stroke="#ea4335" cx="179" cy="90" rx="33" ry="34"></ellipse><ellipse stroke="#fbbc05" cx="270" cy="90" rx="33" ry="34"></ellipse><ellipse cx="358" cy="90" rx="32" ry="34" stroke-width="17px"></ellipse><path d="M391 49l0 73l-2 0a 22 22-30 0 1-59 24"></path><path d="M423 5l0 130" style="stroke:#34a853"></path><path stroke="#ea4335" d="M509 110a32 35-3 1 1 3-29M517 76l-63 24" stroke-width="16px"></path></svg></a></div><div class="bz1lBb"><form class="Pg70bf" id="sf"><a class="l" href="/?lr=lang_en&amp;gbv=1&amp;hl=en&amp;output=search&amp;tbs=lr:lang_1en&amp;sa=X&amp;ved=0ahUKEwix7veuiuHnAhVSDuwKHV6WCPUQPAgE"><svg stroke="#4285f4" aria-label="Google" fill="none" height="22" stroke-width="18" viewbox="0 0 520 174" width="66"><ellipse stroke="#ea4335" cx="179" cy="90" rx="33" ry="34"></ellipse><ellipse stroke="#fbbc05" cx="270" cy="90" rx="33" ry="34"></ellipse><ellipse cx="359" cy="90" rx="31" ry="34"></ellipse><path d="M69 70l61 0m-9 4a56 57 0 1 1-17-47M391 49l0 72l-2 0a 16 17 24 0 1-58 24"></path><path stroke="#34a853" d="M423 5l0 130"></path><path stroke="#ea4335" d="M509 110a32 35-3 1 1 3-29M517 76l-63 24"></path></svg></a><input name="lr" value="lang_en" type="hidden"><input name="gbv" value="1" type="hidden"><input name="hl" value="en" type="hidden"><input name="tbs" value="lr:lang_1en" type="hidden"><div class="H0PQec"><div class="sbc esbc"><input class="noHIxc" value="test" autocapitalize="none" autocomplete="off" name="q" spellcheck="false" type="text"><input name="oq" type="hidden"><input name="aqs" type="hidden"><div class="sc"></div></div></div><button id="qdClwb" type="submit"></button></form></div></header><div id="main"><div><div class="KP7LCb"> <div class="bRsWnc"> <div class="N6RWV"> <div class="Pg70bf Uv67qb"> <span class="OXXup">All</span><a class="eZt8xd" href="/search?q=test&amp;lr=lang_en&amp;gbv=1&amp;hl=en&amp;tbs=lr:lang_1en&amp;source=lnms&amp;tbm=nws&amp;sa=X&amp;ved=0ahUKEwix7veuiuHnAhVSDuwKHV6WCPUQ_AUICCgB">News</a><a class="eZt8xd" href="/search?q=test&amp;lr=lang_en&amp;gbv=1&amp;hl=en&amp;tbs=lr:lang_1en&amp;source=lnms&amp;tbm=isch&amp;sa=X&amp;ved=0ahUKEwix7veuiuHnAhVSDuwKHV6WCPUQ_AUICSgC">Images</a><a class="eZt8xd" href="https://maps.go
---snip---
root@vps:~# curl --http2 -I -A "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36" "https://www.google.com/search?q=test&start=0&gws_rd=cr&gbv=1&lr=lang_en&hl=en&ei=x"
HTTP/2 200 
content-type: text/html; charset=UTF-8
date: Thu, 20 Feb 2020 21:42:26 GMT
expires: -1
cache-control: private, max-age=0
p3p: CP="This is not a P3P policy! See g.co/p3phelp for more info."
server: gws
x-xss-protection: 0
x-frame-options: SAMEORIGIN
expires=Fri, 21-Aug-2020 21:42:26 GMT; path=/; domain=.google.com; Secure; HttpOnly; SameSite=none
alt-svc: quic=":443"; ma=2592000; v="46,43",h3-Q050=":443"; ma=2592000,h3-Q049=":443"; ma=2592000,h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000
accept-ranges: none
vary: Accept-Encoding
```

I guess the reason why Google is blocking Searx really quick is that Searx make requests as if it was Firefox but Firefox actually support by default HTTP2: https://github.com/asciimoo/searx/blob/master/searx/data/useragents.json

If Searx were able to do requests in HTTP2 then it would be like **really** less blocked by Google. ~~It seems that request doesn't support HTTP2 but [Hyper](https://hyper.readthedocs.io/en/latest/) do.~~

**EDIT**: According to this note it seems that request can do HTTP2 requests with hyper as an adapter: https://github.com/khanhicetea/today-i-learned/blob/master/python/HTTP2-supported-for-python-requests-library.md
I was able to make a dirty modification in the file poolrequests.py and it worked:
```patch
1a2
> from hyper.contrib import HTTP20Adapter
68c69
<             self.mount('https://', next(https_adapters))
---
>             self.mount('https://', HTTP20Adapter())
```
Maybe a Python dev could guide me to how to include `HTTP20Adapter()` properly in the code without removing `next(https_adapters)`? (@dalf maybe?)

**Comment by dalf** [comment]
HTTP2 would help also because there is less disconnection after a request, so on the next request there is TCP dance, TLS opening.

Some implementations:
* I haven't tried HTTP20Adapter but [Hyper](https://github.com/python-hyper/hyper) has not been updated since 2017.
* [pycurl](https://github.com/pycurl/pycurl): not sure what to think of this: https://github.com/asciimoo/searx/pull/1725/files#diff-f873ca5d750fe1d8ca82baef73c8e3f3R16 otherwise it works (actually use nghttp2). Main problem: curl and nghttp2 have many different versions across the distro.
* out of topic but related to curl: [nghttp2](https://github.com/nghttp2/nghttp2) has Python bindings only in [server mode](https://nghttp2.org/documentation/python-apiref.html#http-2-servers), not client.
* [aiohttp](https://github.com/aio-libs/aiohttp/tree/master/aiohttp)
   * No native support for HTTP2: see also https://github.com/aio-libs/aiohttp/issues/863
   * [aioh2](https://github.com/decentfox/aioh2): not updated since 2018.
* [hyper-h2](https://github.com/python-hyper/hyper-h2), here the light: it has been updated recently. 
   * Some example: https://python-hyper.org/projects/h2/en/stable/examples.html#example-clients
   * It's only a state machine, to have the whole thing there is [httpx](https://github.com/encode/httpx). (actually use by searx-stats2).
   * HTTP2 is available only using async: a wrapper can be done (but we need to drop Python 2 support before).
   * See: https://github.com/encode/httpx/labels/http%2F2

----

Out of topic: I wish it would be easier to use some of the Golang feature in Python, like the network stack.

**Comment by unixfox** [comment]
Thanks for the reply.

Temporally we could implement HTTP20Adapter as an adapter for request into Searx? Then open another PR for adding HTTP2 support with a maintained library like httpx?

Do you have any idea for my previous question:
> Maybe a Python dev could guide me to how to include HTTP20Adapter() properly in the code without removing next(https_adapters)? (@dalf maybe?)

**Comment by dalf** [comment]
> Maybe a Python dev could guide me to how to include HTTP20Adapter() properly in the code without removing next(https_adapters)? (@dalf maybe?)

You need one HTTP20Adapter per source IP. But [HTTP20Adapter](https://github.com/python-hyper/hyper/blob/development/hyper/contrib.py#L31-L34) seems to ignore the [source_address](https://github.com/asciimoo/searx/blob/master/searx/poolrequests.py#L50) parameter.

So basically HTTP20Adapter needs to be forked:
* to store source_ip
* create a [HTTPConnection](https://github.com/python-hyper/hyper/blob/18b629b8487169870235fe13387e2ae03c178b9f/hyper/common/connection.py#L14) with source_ip.
* Also note the usage of SSL not TLS: https://github.com/python-hyper/hyper/blob/18b629b8487169870235fe13387e2ae03c178b9f/hyper/tls.py#L108
* [EDIT] See also https://github.com/python-hyper/hyper/issues/422

So most probably the whole project need to be forked.

May be I misread the code, I will look more in detail tomorrow or this WE.

**Comment by unixfox** [comment]
Thanks @dalf. I changed the code by moving HTTP20Adapter into the cycle method because I don't use the source IP parameter from `settings.yml`:
```patch
1a2
> from hyper.contrib import HTTP20Adapter
57c58
<     https_adapters = cycle((HTTPAdapterWithConnParams(pool_connections=connect, pool_maxsize=maxsize), ))
---
>     https_adapters = cycle((HTTP20Adapter(pool_connections=connect, pool_maxsize=maxsize), ))
```

The issue tho is that if two requests are made to Searx at the same time then the engine will timeout. I don't really know the reason, if it's due to the behavior of HTTP2 or not. Or maybe I completely don't understand how and where to correctly put the `HTTP20Adapter` in the code.

**Comment by dalf** [comment]
@unixfox I've answered in the issue #899

**Comment by unixfox** [comment]
New update. Google ratelimit less than HTTP1 on HTTP3/QUIC but still ratelimit enough to be blocked after a couple of minutes.

**Comment by djt3** [comment]
My instance is very low traffic so I don't get blacklisted as often, but I ended up making a script that detects if there's a captcha and changes the ip the requests are being made from. This is working well for the past couple of days.

**Comment by return42** [comment]
> i have the problem that google gives me captchas. 

This is 99% due to the fact that bots overwhelm your instance with requests and these requests in form and frequency lead to your instance being blocked by Google.  If you place your instance public to the internet you should really consider to install a filtron reverse proxy and for privacy a result proxy is mandatory.

Please use our new installation instruction and install filtron and morty also.

**Comment by unixfox** [comment]
@return42 why closing this issue? It's still an issue even if we won't see any day soon the solution to the problem.

**Comment by return42** [comment]
@unixfox sorry for my late reply (I have overlooked your comment) ... I closed this issue since I do no longer have this problem since I use filtron.  The docker image and the new [installation description](https://asciimoo.github.io/searx/admin/installation.html) are including filtron. Using these recommended installation methods should solve the problem for others also, so I closed the issue.

If you see reasons to left this issue open, feel free to reopen / from my view the issue is solved.

**Comment by unixfox** [comment]
It highly depends on the popularity of the Searx instance, your instance doesn't receive that many visitors to have Google blocking your IP.
Filtron is just a way to reduce the amount visitors by filtering most of the bots but if you still have too many requests to Google even from good visitors then Google will block your IP. That's the case of my instance and that's why I'm using some proxies.

Just because the problem is solved for you doesn't mean that it's solved for everyone else.

If we had a solution against this Google reCAPTCHA issue directly into Searx then I would call this issue solved but that's not the case [contrary to Invidious](https://github.com/omarroth/invidious/issues/886#issuecomment-552237187).

In conclusion, I would like to have this issue reopened because it would reduce the amount duplicates of this issue [like this one that popped up yesterday](https://github.com/asciimoo/searx/issues/2040).

**Comment by return42** [comment]
> but if you still have too many requests to Google even from good visitors then Google will block your IP. That's the case of my instance and that's why I'm using some proxies.

Sorry, this was not clear to me. / I reopen the issue.

I thought google detects bots when certain search queries come in a certain regularity from the same IP and with the same headers.

**Comment by BBaoVanC** [comment]
I'm using Filtron on my instance and I'm still getting blocked. There's only a handful of users that use my instance. I tried using a tunnel to my server to do the captcha, but it doesn't give me a captcha. Only Searx is given the captcha.

**Comment by dalf** [comment]
@BBaoVanC 
There are multiple criterias to trigger the captcha:
* are you logged (or have been logged) with your Google account ?
* does your browser support Javascript ?
* what are the HTTP headers ?
* does the IP belong to a set use by bot ?
* and most probably more things

If you want to simulate manually something closer to the searx behavior, you can:
* [disable Javascript](https://support.mozilla.org/en-US/questions/994809) (it is a global setting for all windows, all website).
* for each search, start a new private windows

Most probably the captcha will appear in this condition (at least when you use the same IP than your searx instance).

**Comment by BBaoVanC** [comment]
> @BBaoVanC 
> 
> There are multiple criterias to trigger the captcha:
> 
> * are you logged (or have been logged) with your Google account ?
> 
> * does your browser support Javascript ?
> 
> * what are the HTTP headers ?
> 
> * does the IP belong to a set use by bot ?
> 
> * and most probably more things
> 
> 
> 
> If you want to simulate manually something closer to the searx behavior, you can:
> 
> * [disable Javascript](https://support.mozilla.org/en-US/questions/994809) (it is a global setting for all windows, all website).
> 
> * for each search, start a new private windows
> 
> 
> 
> Most probably the captcha will appear in this condition (at least when you use the same IP than your searx instance).

I will try this next time. After a couple hours my server's IP got unblocked. When I tried earlier I was using Firefox 81 with JavaScript enabled (but a private window).

**Comment by bjo81** [comment]
Where is your instance hosted? (which provider)

**Comment by BBaoVanC** [comment]
> Where is your instance hosted? (which provider)

I am running it on a Vultr VPS

**Comment by return42** [comment]
> I'm using Filtron on my instance and I'm still getting blocked.

To inspect if filtron is leaky you can use ([see](https://asciimoo.github.io/searx/utils/filtron.sh.html#overview)):

     sudo -H ./utils/filtron.sh inspect service
    
watch the log for a while and see if questionable requests get through.  If so, improve your [/etc/filtron/rules.json](https://asciimoo.github.io/searx/admin/filtron.html#sample-configuration-of-filtron) to block such requests.

This is what I do when my instance get CAPTCHA from google (had 2 issues in the last 6 month).

**Comment by BBaoVanC** [comment]
> > I'm using Filtron on my instance and I'm still getting blocked.
> 
> To inspect if filtron is leaky you can use ([see](https://asciimoo.github.io/searx/utils/filtron.sh.html#overview)):
> 
> ```
>  sudo -H ./utils/filtron.sh inspect service
> ```
> 
> watch the log for a while and see if questionable requests get through. If so, improve your [/etc/filtron/rules.json](https://asciimoo.github.io/searx/admin/filtron.html#sample-configuration-of-filtron) to block such requests.
> 
> This is what I do when my instance get CAPTCHA from google (had 2 issues in the last 6 month).

I will try that, but I don't think that there's many other people using my instance. I checked nginx logs and I only saw myself, with a few other ips occasionally.

**Comment by unixfox** [comment]
@BBaoVanC Google seems to only serve a CAPTCHA on per HTTP protocol. Searx use HTTP1 and your browser is using HTTP2 so that's why you aren't getting the CAPTCHA on Firefox. If you try with curl or any browser in HTTP1 you will see the CAPTCHA.

I know that it's a strange behavior but that's how it works.

I talked about it here: https://github.com/asciimoo/searx/issues/729#issuecomment-589334071

**Comment by BBaoVanC** [comment]
> @BBaoVanC Google seems to only serve a CAPTCHA on per HTTP protocol. Searx use HTTP1 and your browser is using HTTP2 so that's why you aren't getting the CAPTCHA on Firefox. If you try with curl or any browser in HTTP1 you will see the CAPTCHA.
> 
> I know that it's a strange behavior but that's how it works.

Is there a way to get Searx to use HTTP2 requests? Or would that help? Next time I get a CAPTCHA I will also try forcing Firefox to use HTTP1.

**Comment by dalf** [comment]
> Is there a way to get Searx to use HTTP2 requests? Or would that help? Next time I get a CAPTCHA I will also try forcing Firefox to use HTTP1.

If you want HTTP/2 right away, you can use this PR : https://github.com/asciimoo/searx/pull/1856

**Comment by return42** [comment]
> I will try that, but I don't think that there's many other people using my instance. 

>  with a few other ips occasionally.

Same to mine, but when I have a CAPTCHA issue, bots came through with the same questionable request every minute or two .. this seems enough for goolge to block my instance.

Anyway it is very interesting to monitor the log of filtron :)

**Comment by BBaoVanC** [comment]
> > I will try that, but I don't think that there's many other people using my instance.
> 
> > with a few other ips occasionally.
> 
> Same to mine, but when I have a CAPTCHA issue, bots came through with the same questionable request every minute or two .. this seems enough for goolge to block my instance.
> 
> Anyway it is very interesting to monitor the log of filtron :)

It looks like the same log is also shown with journalctl on the filtron service, which is what I was using a while ago.

**Comment by BBaoVanC** [comment]
I got blocked again today, and I switched my User Agent to the one used by Searx. I also disabled Javascript, and disabled cookies. Google still let me through while blocking my Searx instance.

**Comment by minami-o** [comment]
Whenever my instance was blocked by Google, I found out that my filtron was not active or set up properly.
If you use a web server as reverse proxy, check its logs for spam queries and adapt your filtron rules in case you find any.
Minami-o

**Comment by BBaoVanC** [comment]
> Whenever my instance was blocked by Google, I found out that my filtron was not active or set up properly. If you use a web server as reverse proxy, check its logs for spam queries and adapt your filtron rules in case you find any. Minami-o Le 5 août 2020 15:46:26 GMT-04:00, BBaoVanC <notifications@github.com> a écrit :
> […](#)
> I got blocked again today, and I switched my User Agent to the one used by Searx. I also disabled Javascript, and disabled cookies. Google still let me through while blocking my Searx instance. -- You are receiving this because you are subscribed to this thread. Reply to this email directly or view it on GitHub: [#729 (comment)](https://github.com/asciimoo/searx/issues/729#issuecomment-669455739)

I've actually already done this. In the nginx logs for Searx, I only see requests from me, and a couple every so often from other people.

**Comment by unixfox** [comment]
> I got blocked again today, and I switched my User Agent to the one used by Searx. I also disabled Javascript, and disabled cookies. Google still let me through while blocking my Searx instance.

You tried in HTTP2 or HTTP1?

**Comment by BBaoVanC** [comment]
> > I got blocked again today, and I switched my User Agent to the one used by Searx. I also disabled Javascript, and disabled cookies. Google still let me through while blocking my Searx instance.
> 
> You tried in HTTP2 or HTTP1?

It looks like I didn't. I checked about:config in firefox and http.version was set to 1.1 by default, but the developer console says it's using HTTP2

**Comment by unixfox** [comment]
> > > I got blocked again today, and I switched my User Agent to the one used by Searx. I also disabled Javascript, and disabled cookies. Google still let me through while blocking my Searx instance.
> > 
> > 
> > You tried in HTTP2 or HTTP1?
> 
> It looks like I didn't. I checked about:config in firefox and http.version was set to 1.1 by default, but the developer console says it's using HTTP2

Then try again but with curl:
```
curl --http1.1 https://www.google.com/search?q=test
```
Let me know if you see a CAPTCHA in the raw HTML code.

**Comment by ltguillaume** [comment]
Could someone explain to me why sometimes the entire Searx results page is being blocked by the captcha error (it _only_ shows `Engines cannot retrieve results:
google (unexpected crash: CAPTCHA required)`) while at other times, a similar message is put into the right column and results from other engines are still being shown? Obviously the latter is the preferred behavior.

**Comment by unixfox** [comment]
@ltGuillaume it could be that the other engine are also failing. I don't know which engine you activated so it's difficult to diagnose your issue.

**Comment by ltguillaume** [comment]
> @ltGuillaume it could be that the other engine are also failing. I don't know which engine you activated so it's difficult to diagnose your issue.

Wikipedia, Bing, (Startpage,) Google on v0.17, mostly using https://search.disroot.org. You're probably right: if I disable Google, no results other than Wikipedia's in the sidebar show up. So all these engines fail now?

**Comment by unixfox** [comment]
> Wikipedia, Bing, (Startpage,) Google on v0.17, mostly using [search.disroot.org](https://search.disroot.org). You're probably right: if I disable Google, no results other than Wikipedia's in the sidebar show up. So all these engines fail now?

Not all but Bing sometimes fail, related issue here: https://github.com/asciimoo/searx/issues/2019

**Comment by NightMachinery** [comment]
> what about forwarding the captcha to the frontend and save the cookie in the backend then?

Can we have a separate issue tracking this? It's the only robust solution for Google. Google required a captcha for my VPS from the very start (its IP was probably blacklisted or sth).

**Comment by immanuelfodor** [comment]
VPS providers are often blocked by other services as well using their complete IP block or ASN number. People often use VPSes for webscraping or sending spam, so I wouldn't be surprised if this is the case why you got a captcha from the start.

**Comment by deepend-tildeclub** [comment]
Does duckduckgo have any of these issues?   Maybe design searx to grab google results through duckduckgo.. since duckduckgo does offer google search.

**Comment by unixfox** [comment]
> Does duckduckgo have any of these issues?   Maybe design searx to grab google results through duckduckgo.. since duckduckgo does offer google search. 

Duckduckgo uses mostly Bing results and sometimes Google. From my experience the results quality on Duckduckgo is really bad compared to Google.

Why would we fetch Duckduckgo results for the Google engine if there is already a Duckduckgo engine? That doesn't make any sense, just disable the Google engine and use the Duckduckgo engine if you want to do that.

**Comment by deepend-tildeclub** [comment]
> > Does duckduckgo have any of these issues?   Maybe design searx to grab google results through duckduckgo.. since duckduckgo does offer google search.
> 
> Duckduckgo uses mostly Bing results and sometimes Google. From my experience the results quality on Duckduckgo is really bad compared to Google.
> 
> Why would we fetch Duckduckgo results for the Google engine if there is already a Duckduckgo engine? That doesn't make any sense, just disable the Google engine and use the Duckduckgo engine if you want to do that.

for the people that want google results only and this captcha crap is causing issues.  you can use duckduckgo for searching only google if you want.   Wonder how they get those results without the same issues.

**Comment by BBaoVanC** [comment]
> > > Does duckduckgo have any of these issues?   Maybe design searx to grab google results through duckduckgo.. since duckduckgo does offer google search.
> > 
> > 
> > Duckduckgo uses mostly Bing results and sometimes Google. From my experience the results quality on Duckduckgo is really bad compared to Google.
> > Why would we fetch Duckduckgo results for the Google engine if there is already a Duckduckgo engine? That doesn't make any sense, just disable the Google engine and use the Duckduckgo engine if you want to do that.
> 
> for the people that want google results only and this captcha crap is causing issues. you can use duckduckgo for searching only google if you want. Wonder how they get those results without the same issues.

What do you mean by "you can use duckduckgo for searching only google"?

**Comment by ltguillaume** [comment]
> What do you mean by "you can use duckduckgo for searching only google"?

My thoughts exactly. AFAIK you can only use !g, which simply opens up a Google window...

**Comment by mj162** [comment]
Happy Christmas.

A gentle introduction to what's causing the Google CAPTCHA crap would be appreciated.

The searx instances I tend to use are .ir, .xyx .lnode.net.  I didn't completely follow the discussion but gather Google is refusing based on the searx instance IP.  My first question: Are the searx instances passing the ip of the browser performing the search?  I ask because I've refused to accept Google's terms for using their search service, so always get the result below.  If my IP is being passed, won't it always fail for me?:

`curl --http1.1 https://www.google.com/search?q=test`

`Your client does not have permission to get URL <code>/search?q=test</code> from this server.  (Client IP address: w.x.y.z)`

I get that we're being told to deselect Google from the backend search engines that a searx instance uses, but my Gran wouldn't know how to do that after I've got her to use searx.local instead of Google.  So my second question: Without my Gran having to do anything, cannot searx instances fail-over and try ddg or another backend if Google refuses?

Happy New Year.

**Comment by unixfox** [comment]
> Happy Christmas.
> 
> A gentle introduction to what's causing the Google CAPTCHA crap would be appreciated.
> 
> The searx instances I tend to use are .ir, .xyx .lnode.net.  I didn't completely follow the discussion but gather Google is refusing based on the searx instance IP.  My first question: Are the searx instances passing the ip of the browser performing the search?  I ask because I've refused to accept Google's terms for using their search service, so always get the result below.  If my IP is being passed, won't it always fail for me?:
> 
> `curl --http1.1 https://www.google.com/search?q=test`
> 
> `Your client does not have permission to get URL <code>/search?q=test</code> from this server.  (Client IP address: w.x.y.z)`

Only the server IP is sending the requests to Google servers. That's the reason why Searx is good for privacy. So it doesn't matter if you refused or accepted Google terms.

> I get that we're being told to deselect Google from the backend search engines that a searx instance uses, but my Gran wouldn't know how to do that after I've got her to use searx.local instead of Google.  So my second question: Without my Gran having to do anything, cannot searx instances fail-over and try ddg or another backend if Google refuses?
> 
> Happy New Year.

There is no fail-over mecanism because Searx is a meta search not a search provider so it isn't only dedicated to only one engine (in this case Google).
But you could enable both ddg and Google engines then if Google engine fails you still have ddg results. Do note that this adds a bit more latency and the response time won't be as quick as with only one engine.

If you want a Searx instance that works great with Google you can try https://searx.be

-----

Merry christmas and happy new year.

**Comment by mj162** [comment]
Happy Christmas

I've obviously not understood how searx works.  I _**did**_ understand that searx was not a search provider but I didn't understand whether it sends to several providers in parallel and returns the first to answer or whether it works through a statically-defined or dynamically-updated precedence order.  My Gran just wants to search and is non-technical; and in comparison with the searx developers, I'm probably not much better.

Happy New Year

**Comment by edmundlaugasson** [comment]
Still an error using https://search.privacytools.io/searx/ :
`google (unexpected crash CAPTCHA required)`
This means, that second and rest of search pages won't open....
Is this bug tend to be fixed? I see it exist already since 2016...

**Comment by unixfox** [comment]
> Still an error using [search.privacytools.io/searx](https://search.privacytools.io/searx/) :
> `google (unexpected crash CAPTCHA required)`
> This means, that second and rest of search pages won't open....
> Is this bug tend to be fixed? I see it exist already since 2016...

It's not a bug because this issue is unsolvable by the Searx maintainers due to the fact that Google blocks the requests because it considers Searx as a bot.

Some improvements since 2016 were made for combatting against this issue:
- The ability to use multiple proxies in order to make HTTP requests from multiple IPs: https://github.com/searx/searx/pull/2319
- Trying to make Searx look like a more real browser in https://github.com/searx/searx/pull/1856 (still in progress)
- [Filtron](https://github.com/asciimoo/filtron), a tool for reducing the amount of bot requests on Searx public instances so that less useless requests are made to Google servers.

But most of the job has to be made by the maintainer of the Searx instance (in this case https://search.privacytools.io/searx/) because we offer the tools to combat against this issue but if the maintainer doesn't use these tools then we can't do much about this issue.

You may also use your own Searx instance instead of using public Searx instances, we have an official guide for that here: https://searx.github.io/searx/admin/installation.html, and you won't face this issue because you will be the only one to use it.

Just a self-promotion, I'm running a public instance which is designed to work (most of the time) with Google if you are interested: https://searx.be.

**Comment by return42** [comment]
@unixfox: great explained! I think your response is the ultimate answer of this issue / can we now close this issue with it?

**Comment by unixfox** [comment]
> @unixfox: great explained! I think your response is the ultimate answer of this issue / can we now close this issue with it?
> 

No we won't close this issue because I have a feeling that closing it will create duplicate of it because people won't find this issue and thus think of creating a new one.

Technically we still have some work on our side for combatting this issue, an example is https://github.com/searx/searx/issues/2439

**Comment by ltGuillaume** [comment]
> Just a self-promotion, I'm running a public instance which is designed to work (most of the time) with Google if you are interested: https://searx.be.
@unixfox So according to https://searx.space you're running this on a US server? 😟

**Comment by unixfox** [comment]
> > Just a self-promotion, I'm running a public instance which is designed to work (most of the time) with Google if you are interested: [searx.be](https://searx.be).
> > @unixfox So according to [searx.space](https://searx.space) you're running this on a US server? worried

No I'm not, it's just the company that is US based, but it's actually hosted in the Netherlands.

**Comment by ltGuillaume** [comment]
But if the company is US-based, US laws apply to your instance? (I wonder if that entails much when it comes to a searx instance, but it does feel less trustworthy).

**Comment by mj162** [comment]
Agree with the logic in keeping this ticket open.  It's how I found it before otherwise creating a new ticket.

I have switched to Unixfox's instance as first preference (though oddly, language has to be set to English UK rather than US for search results to find anything -- seen that on other instances too).

**Comment by deepend-tildeclub** [comment]
Not speaking from experience because I have none.  But is there a way to proxy the Google captcha and show it to the client to respond.  Or is that impossible? 

Don't think making Searx look more like a real browser would solve anything because it's the amount of requests from one ip that causes Google to ask for captcha.   But maybe having more ips and have searx spread the requests over multiple ips as well as blocking bot requests could lessen the issue a bit.

**Comment by unixfox** [comment]
> Not speaking from experience because I have none.  But is there a way to proxy the Google captcha and show it to the client to respond.  Or is that impossible? 

This has been brought up many times and no it's impossible because a recaptcha can only be loaded from the same domain, in this case www.google.com

> I have switched to Unixfox's instance as first preference (though oddly, language has to be set to English UK rather than US for search results to find anything -- seen that on other instances too).

Please open an issue if you can replicate the bug on other instances and even in a local Searx.

**Comment by mj162** [comment]
I've used this: [text](https://pastebin.com/LNZHzp9S) proxy for years.  The gentleman moved to a different hosting provider and since then Google's baulked with its graphical Captcha request...  Agreed, too many requests from one IP is a reasonable assumption but how many is too many and how far back in time are they going; the last 5 minutes or the last 5 years?

[will do]

**Comment by paradoxsupreme** [comment]
I am trying to self host a searx instance on a VPS with nginx. I have used [these](https://github.com/searx/searx/blob/master/utils/templates/etc/filtron/rules.json) filtron rules.json. I have also set up morty. The instance is working correctly but when I try to use the google search engine it gives me ` google (too many requests) ` on the first request and then it gives `Sorry! we didn't find any results.`. Is there anything I can do to fix the google search egine?

**Comment by unixfox** [comment]
> I am trying to self host a searx instance on a VPS with nginx. I have used [these](https://github.com/searx/searx/blob/master/utils/templates/etc/filtron/rules.json?rgh-link-date=2021-02-16T12%3A33%3A53Z) filtron rules.json. I have also set up morty. The instance is working correctly but when I try to use the google search engine it gives me `google (too many requests)` on the first request and then it gives `Sorry! we didn't find any results.`. Is there anything I can do to fix the google search egine?

You could use some proxies but unfortunately apart from that Searx can't do much about this issue.

**Comment by paradoxsupreme** [comment]
> > I am trying to self host a searx instance on a VPS with nginx. I have used [these](https://github.com/searx/searx/blob/master/utils/templates/etc/filtron/rules.json?rgh-link-date=2021-02-16T12%3A33%3A53Z) filtron rules.json. I have also set up morty. The instance is working correctly but when I try to use the google search engine it gives me `google (too many requests)` on the first request and then it gives `Sorry! we didn't find any results.`. Is there anything I can do to fix the google search egine?
> 
> You could use some proxies but unfortunately apart from that Searx can't do much about this issue.

Could you give me some resources about how I can configure these proxies?

**Comment by unixfox** [comment]
> > > I am trying to self host a searx instance on a VPS with nginx. I have used [these](https://github.com/searx/searx/blob/master/utils/templates/etc/filtron/rules.json?rgh-link-date=2021-02-16T14%3A00%3A40Z) filtron rules.json. I have also set up morty. The instance is working correctly but when I try to use the google search engine it gives me `google (too many requests)` on the first request and then it gives `Sorry! we didn't find any results.`. Is there anything I can do to fix the google search egine?
> > 
> > 
> > You could use some proxies but unfortunately apart from that Searx can't do much about this issue.
> 
> Could you give me some resources about how I can configure these proxies?

Add them into the settings.yaml here: https://github.com/searx/searx/blob/master/searx/settings.yml#L78

**Comment by paradoxsupreme** [comment]
> > > > I am trying to self host a searx instance on a VPS with nginx. I have used [these](https://github.com/searx/searx/blob/master/utils/templates/etc/filtron/rules.json?rgh-link-date=2021-02-16T14%3A00%3A40Z) filtron rules.json. I have also set up morty. The instance is working correctly but when I try to use the google search engine it gives me `google (too many requests)` on the first request and then it gives `Sorry! we didn't find any results.`. Is there anything I can do to fix the google search egine?
> > > 
> > > 
> > > You could use some proxies but unfortunately apart from that Searx can't do much about this issue.
> > 
> > 
> > Could you give me some resources about how I can configure these proxies?
> 
> Add them into the settings.yaml here: https://github.com/searx/searx/blob/master/searx/settings.yml#L78

Thanks

**Comment by paradoxsupreme** [comment]
> > > > I am trying to self host a searx instance on a VPS with nginx. I have used [these](https://github.com/searx/searx/blob/master/utils/templates/etc/filtron/rules.json?rgh-link-date=2021-02-16T14%3A00%3A40Z) filtron rules.json. I have also set up morty. The instance is working correctly but when I try to use the google search engine it gives me `google (too many requests)` on the first request and then it gives `Sorry! we didn't find any results.`. Is there anything I can do to fix the google search egine?
> > > 
> > > 
> > > You could use some proxies but unfortunately apart from that Searx can't do much about this issue.
> > 
> > 
> > Could you give me some resources about how I can configure these proxies?
> 
> Add them into the settings.yaml here: https://github.com/searx/searx/blob/master/searx/settings.yml#L78

I tried to uncomment these lines to settings.yml but I get internal server error.

**Comment by unixfox** [comment]
> > > > > I am trying to self host a searx instance on a VPS with nginx. I have used [these](https://github.com/searx/searx/blob/master/utils/templates/etc/filtron/rules.json?rgh-link-date=2021-02-16T14%3A59%3A03Z) filtron rules.json. I have also set up morty. The instance is working correctly but when I try to use the google search engine it gives me `google (too many requests)` on the first request and then it gives `Sorry! we didn't find any results.`. Is there anything I can do to fix the google search egine?
> > > > 
> > > > 
> > > > You could use some proxies but unfortunately apart from that Searx can't do much about this issue.
> > > 
> > > 
> > > Could you give me some resources about how I can configure these proxies?
> > 
> > 
> > Add them into the settings.yaml here: [https://github.com/searx/searx/blob/master/searx/settings.yml#L78](https://github.com/searx/searx/blob/master/searx/settings.yml?rgh-link-date=2021-02-16T14%3A59%3A03Z#L78)
> 
> I tried to uncomment these lines to settings.yml but I get internal server error

Please open a discussion in this category: https://github.com/searx/searx/discussions/categories/installation-issue-custom with your settings.yaml.

**Comment by BBaoVanC** [comment]
> > I am trying to self host a searx instance on a VPS with nginx. I have used [these](https://github.com/searx/searx/blob/master/utils/templates/etc/filtron/rules.json?rgh-link-date=2021-02-16T12%3A33%3A53Z) filtron rules.json. I have also set up morty. The instance is working correctly but when I try to use the google search engine it gives me `google (too many requests)` on the first request and then it gives `Sorry! we didn't find any results.`. Is there anything I can do to fix the google search egine?
> 
> You could use some proxies but unfortunately apart from that Searx can't do much about this issue.

Where do I get proxies?

**Comment by unixfox** [comment]
> > > I am trying to self host a searx instance on a VPS with nginx. I have used [these](https://github.com/searx/searx/blob/master/utils/templates/etc/filtron/rules.json?rgh-link-date=2021-02-16T12%3A33%3A53Z) filtron rules.json. I have also set up morty. The instance is working correctly but when I try to use the google search engine it gives me `google (too many requests)` on the first request and then it gives `Sorry! we didn't find any results.`. Is there anything I can do to fix the google search egine?
> > 
> > You could use some proxies but unfortunately apart from that Searx can't do much about this issue.
> 
> Where do I get proxies?

Please consult this discussion where I already answered to this question: https://github.com/searx/searx/discussions/2576#discussioncomment-373973

**Comment by rob4226** [comment]
Hi, I've been using searx for a long time and love it, I just updated to v1.0.  I don't think this is unique to v1.0 but I keep getting this error for google that many others get:
```bash
Engines cannot retrieve results: google (too many requests)
```
I understand that means too many requests from one IP but what I don't understand is that both my self-hosted searx server and my computer are on the same network in my house and hence use the same external IP address...but when I go to google in my browser it searches fine, no CAPTCHA, or too many requests error.  

**So why would I be getting a too many requests error from google when using searx, but not my browser, when coming from the same exact IP?**  Just doesn't make sense to me.

_Thank you for everyone's work on searx!_

**Comment by unixfox** [comment]
> Hi, I've been using searx for a long time and love it, I just updated to v1.0. I don't think this is unique to v1.0 but I keep getting this error for google that many others get:
> 
> ```shell
> Engines cannot retrieve results: google (too many requests)
> ```
> 
> I understand that means too many requests from one IP but what I don't understand is that both my self-hosted searx server and my computer are on the same network in my house and hence use the same external IP address...but when I go to google in my browser it searches fine, no CAPTCHA, or too many requests error.
> 
> **So why would I be getting a too many requests error from google when using searx, but not my browser, when coming from the same exact IP?** Just doesn't make sense to me.
> 
> _Thank you for everyone's work on searx!_

Please see https://github.com/searx/searx/issues/729#issuecomment-589334071 and https://github.com/searx/searx/issues/729#issuecomment-668447548.

**Comment by RennisDitchie** [comment]
@unixfox 
Hey there, thanks for the post above which explains that the overarching issue isn't anything to deal with Searx directly, but in that Google treats Searx as a bot (go figure, botnet search engine doesn't like other bots...)

In this case, how would you go about DISABLING Google from a Searx instance, and just use DuckDuckGo instead?

I ask because I am running into the same exact "Engines cannot retrieve results: google (too many requests)" issue as everyone else since I host it on a Debian 10 VPS droplet on Digital Ocean.

I was willing to bet that maybe I did too many search requests, but it'll often give this error on literally the first search attempt if I include too many words in the query itself.

Anyway, would appreciate if you could link me to a specific section of the guide on how to do this, or another issue as I just want to nuke Google from my Searx config to just get rid of the issue entirely as I can't even use the stupid search engine as a result of the error.

Thanks again.

**Comment by BBaoVanC** [comment]
On 04/23/21 10:19PM, RennisDitchie wrote:
> @unixfox 
> Hey there, thanks for the post above which explains that the overarching issue isn't anything to deal with Searx directly, but in that Google treats Searx as a bot (go figure, botnet search engine doesn't like other bots...)
> 
> In this case, how would you go about DISABLING Google from a Searx instance, and just use DuckDuckGo instead?
> 
> I ask because I am running into the same exact "Engines cannot retrieve results: google (too many requests)" issue as everyone else since I host it on a Debian 10 VPS droplet on Digital Ocean.
> 
> I was willing to bet that maybe I did too many search requests, but it'll often give this error on literally the first search attempt if I include too many words in the query itself.
> 
> Anyway, would appreciate if you could link me to a specific section of the guide on how to do this, or another issue as I just want to nuke Google from my Searx config to just get rid of the issue entirely as I can't even use the stupid search engine as a result of the error.
> 
> Thanks again.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/searx/searx/issues/729#issuecomment-826037086

Go to your settings.yml and set `disabled` to `True` under the `google`
engine.

Ex.

```yml
  - name : google
    engine : google
    shortcut : go
    disabled : True
```

You can also completely delete it (so it doesn't show up at all) by just removing this code block from your settings.yml.

**Comment by Pofilo** [comment]
@RennisDitchie 

> I can't even use the stupid search engine

Well, you could be more respectful from something you can use/modify freely right ?

However, this Google issue is something we can't really resolve.
Another metasearch engine named [Whoogle](https://github.com/benbusby/whoogle-search) is designed only for Google results and suffers from the same problems (I tried it last week).

**Comment by RennisDitchie** [comment]
@Pofilo 

I am thankful, but when I first deployed Searx a year ago just months after Luke Smith's video on the same topic, it just didn't work period, plus the installation guide was a mess.

Yes, the deployment process has improved, and yes, I can solve my issues by "using the docker image".

Don't confuse my frustration for the fact that I think your guys' project is awesome.

I'm actually more mad at Google itself for blocking it in the first place. Bear with me on that one since I'm just debating how to fix the issue itself. I thought it was my newbness in terms of not following the instructions to a T, but am kind of glad its really just an issue that will most likely never be truly solved since Google will always try to circumvent users from not directly accessing their website.

Anyway, just wanted to say thanks on that regard.

**The only other issue I'm running into is that on basic queries like "text outline css", I get this error as well:**
`Sorry! we didn't find any results. Please use another query or search in more categories. `

I'm assuming I have to check the error log to see why this is occurring?

**Comment by RennisDitchie** [comment]
> On 04/23/21 10:19PM, RennisDitchie wrote: @unixfox Hey there, thanks for the post above which explains that the overarching issue isn't anything to deal with Searx directly, but in that Google treats Searx as a bot (go figure, botnet search engine doesn't like other bots...) In this case, how would you go about DISABLING Google from a Searx instance, and just use DuckDuckGo instead? I ask because I am running into the same exact "Engines cannot retrieve results: google (too many requests)" issue as everyone else since I host it on a Debian 10 VPS droplet on Digital Ocean. I was willing to bet that maybe I did too many search requests, but it'll often give this error on literally the first search attempt if I include too many words in the query itself. Anyway, would appreciate if you could link me to a specific section of the guide on how to do this, or another issue as I just want to nuke Google from my Searx config to just get rid of the issue entirely as I can't even use the stupid search engine as a result of the error. Thanks again. -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#729 (comment)](https://github.com/searx/searx/issues/729#issuecomment-826037086)
> Go to your settings.yml and set `disabled` to `True` under the `google` engine. Ex. ```yml - name : google engine : google shortcut : go disabled : True ``` You can also completely delete it (so it doesn't show up at all) by just removing this code block from your settings.yml.

Thanks a ton for this!

Will write this down in my org doc notes, thanks!

**Comment by RennisDitchie** [comment]
After disabling Google, I'm now running into the "Sorry! we didn't find any results. Please use another query or search in more categories." error more often.

However, I will note this within the relevant GitHub issue instead:
https://github.com/searx/searx/issues/2053

Thank you again to both @BBaoVanC @Pofilo and @unixfox for providing great insight on this issue within this GitHub issue.

I know I might have seemed a bit harsh in my responses before, and trust me, I'm just not as big brained to create an entire Python based project as Searx on my own, so I just wanted to say thanks for the help on this one.
